hw2/hw2.py

```diff
@@ -0,0 +1,420 @@
+        """
+        self.children.append(node)
+        self.children_values.append(val)
+
+    def calc_feature_importance(self, n_total_sample):
+        """
+        Calculate the selected feature importance.
+
+        Input:
+        - n_total_sample: the number of samples in the dataset.
+
+        This function has no return value - it stores the feature importance in
+        self.feature_importance
+        """
+        if (
+            self.terminal or self.feature == -1
+        ):  # Terminal or no feature used for splitting
+            self.feature_importance = 0
+            return
+
+        # Impurity of the current node
+        n_samples = self.data.shape[0]  # Number of samples at the current node
+        node_impurity = self.impurity_func(self.data)
+
+        # Calculate weighted impurity for the current node
+        weighted_impurity_current = (n_samples / n_total_sample) * node_impurity
+
+        # Calculate the sum of weighted impurities for each child node
+        weighted_impurities_children = np.sum(
+            [
+                (child.data.shape[0] / n_total_sample) * child.impurity_func(child.data)
+                for child in self.children
+            ]
+        )
+
+        # Feature importance calculation
+        self.feature_importance = (
+            weighted_impurity_current - weighted_impurities_children
+        )
+
+    def goodness_of_split(self, feature, gain_ratio=False):
+        """
+        Calculate the goodness of split of a dataset given a feature and impurity function.
+
+        Input:
+        - feature: the feature index the split is being evaluated according to.
+
+        Returns:
+        - goodness: the goodness of split
+        - groups: a dictionary holding the data after splitting
+                  according to the feature values.
+        """
+        # Splitting the data by feature values
+        unique_values, indices = np.unique(self.data[:, feature], return_inverse=True)
+        groups = {
+            val: self.data[indices == idx] for idx, val in enumerate(unique_values)
+        }
+
+        # Calculate impurity before the split
+        impurity_before = self.impurity_func(self.data)
+
+        # Calculate weighted impurity after the split
+        n_samples = len(self.data)
+        weighted_impurity_after = sum(
+            (len(group) / n_samples) * self.impurity_func(group)
+            for group in groups.values()
+        )
+
+        # Goodness of Split
+        goodness = impurity_before - weighted_impurity_after
+
+        if gain_ratio:
+            # Calculate split information
+            split_info = -sum(
+                (len(group) / n_samples) * np.log2(len(group) / n_samples)
+                for group in groups.values()
+                if len(group) > 0  # Avoid log2(0) which is undefined
+            )
+            # Gain Ratio
+            if split_info > 0:
+                goodness = goodness / split_info
```

Comment:
By definition, gain ratio uses information-gain.
Meaning, entropy in calculations.
Thus, we should use it explicitly and not build on `self.impurity_func == entropy`.

Reply:
Good point!

Reply:
Updated